### PR TITLE
Enable chemist-friendly depiction of R-groups

### PR DESCRIPTION
**Pull request by ptosco**
This is my second attempt at #6746, this time following @greglandrum suggestions (it was actually a bit stronger than a suggestion ;-) ).
Let's see how it goes this time.

- added a convenience function to relabel R-groups in a chemist-friendly fashion
- exposed functionality to MinimalLib through a JSON option at molecule creation time
- added relevant unit tests

### CODE REVIEW

File: Code/GraphMol/RGroupDecomposition/RGroupUtils.h [hunk 1]
```diff
@@ -93,6 +93,23 @@ RDKIT_RGROUPDECOMPOSITION_EXPORT std::string toJSON(
...
+/// than #*, *:#, #*:#, etc. By default, only the MDLRGroup label
+/// is retained on output; this may be configured through the
+/// outputLabels parameter.
+/// In case there are multiple potential R-group mappings,
+/// the priority on input is Atom map number > Isotope > MDLRGroup.
+/// The inputLabels parameter allows to configure which mappings
+/// are taken into consideration.
+RDKIT_RGROUPDECOMPOSITION_EXPORT void relabelMappedDummies(
+    ROMol &mol,
+    RGroupLabelling inputLabels = static_cast<RGroupLabelling>(AtomMap |
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupUtils.h:108]: having to include the static_cast here is pretty ugly.
I think I'd prefer that both this parameter and `outputLabels` be unsigned ints, like the `labels` and `rgroupLabeling` members of `RGroupDecompositionParameters`.
What's your take on that?

File: Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp [hunk 2]
```diff
@@ -208,4 +208,35 @@ std::string toJSON(const RGroupColumns &cols, const std::string &prefix) {
...
+    if (!atomMapNum && (inputLabels & MDLRGroup)) {
+      atom->getPropIfPresent(common_properties::_MolFileRLabel, atomMapNum);
+    }
+    if (!atomMapNum) {
+      continue;
+    }
+    auto rLabel = "R" + std::to_string(atomMapNum);
+    atom->setProp(common_properties::dummyLabel, rLabel);
+    atom->setAtomMapNum((outputLabels & AtomMap) ? atomMapNum : 0);
+    atom->setIsotope((outputLabels & Isotope) ? atomMapNum : 0);
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp:233]: I think this is going too far with keeping the code short (we don't have to pay by the line of code ;-) ), how about this:
```suggestion
    if(outputLabels & AtomMap) {
      atom->setAtomMapNum(atomMapNum);
    } else {
      atom->setAtomMapNum(0);
    }
    if(outputLabels & Isotope) {
      atom->setIsotope(atomMapNum);
    } else {
      atom->setIsotope(0);
    }
```

File: Code/GraphMol/RGroupDecomposition/catch_rgd.cpp [hunk 3]
```diff
@@ -808,3 +808,175 @@ TEST_CASE("Mol matches core") {
...
+    CHECK(
+        MolToCXSmiles(*core, p) ==
+        "c1cc([4*:2])c([3*:1])cn1 |atomProp:3.dummyLabel.*:3.molAtomMapNumber.2:5.dummyLabel.*:5.molAtomMapNumber.1|");
+    relabelMappedDummies(*core, Isotope);
+    CHECK(MolToCXSmiles(*core, p) ==
+          "c1cc(*)c(*)cn1 |atomProp:3.dummyLabel.R4:5.dummyLabel.R3|");
+  }
+  SECTION(
+      "AtomMap, Isotope and MDLRGroup in, MDLRGroup out - AtomMap has priority") {
+    auto core = ROMOL_SPTR(new ROMol(*allDifferentCore));
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/catch_rgd.cpp:912]: Definitely don't need a shared pointer here. I think you don't need a pointer at all, why not just:
```suggestion
    ROMol core(*allDifferentCore);
```

same suggestion applies to the other tests below

File: Code/GraphMol/RGroupDecomposition/catch_rgd.cpp [hunk 4]
```diff
@@ -808,3 +808,175 @@ TEST_CASE("Mol matches core") {
...
+          "c1cc(*)c(*)cn1 |atomProp:3.dummyLabel.R2:5.dummyLabel.R1|");
+  }
+  SECTION(
+      "AtomMap, Isotope and MDLRGroup in, AtomMap out - AtomMap has priority") {
+    auto core = ROMOL_SPTR(new ROMol(*allDifferentCore));
+    relabelMappedDummies(
+        *core, static_cast<RGroupLabelling>(AtomMap | Isotope | MDLRGroup),
+        AtomMap);
+    CHECK(
+        MolToCXSmiles(*core, p) ==
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/catch_rgd.cpp:938]: I think the dummy Labels shouldn't still be in the output here?

Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/catch_rgd.cpp:938]: This applies below a few times as well

**Review by greglandrum — CHANGES_REQUESTED** [review]
Some small suggestions

File: Code/GraphMol/RGroupDecomposition/RGroupUtils.h (hunk 1 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupUtils.h:108]: I agree.

File: Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp (hunk 2 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupUtils.cpp:233]: This is part of my "Greg alertness test" campaign: the day you won't flag a ternary conditional operator or >5-line-long logic test I will start worrying. Great that this has not happened yet.
(I still prefer the ternary operator to the above, but that does not matter as much)

File: Code/GraphMol/RGroupDecomposition/catch_rgd.cpp (hunk 3 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/catch_rgd.cpp:912]: I agree, that's unnecessary.

File: Code/GraphMol/RGroupDecomposition/catch_rgd.cpp (hunk 4 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/catch_rgd.cpp:938]: Ah, good point. Fixed.

**Comment by ptosco** [comment]
@greglandrum I should have addressed all your poiints.

**Comment by greglandrum** [comment]
@ptosco  : the test failures are real

**Comment by ptosco** [comment]
> @ptosco : the test failures are real

Yeah, I removed the `dummyLabel` from the C++ test but I had forgotten about the Python one; should be good now.

**Review by greglandrum — APPROVED** [review]
LGTM

**Pull request merged**